diff --git a/cmake/HPX_SetupCUDA.cmake b/cmake/HPX_SetupCUDA.cmake
index 1c881f9e68..8e9a538bd7 100644
--- a/cmake/HPX_SetupCUDA.cmake
+++ b/cmake/HPX_SetupCUDA.cmake
@@ -62,5 +62,5 @@ if(HPX_WITH_CUDA AND NOT TARGET hpx::cuda)
     target_link_libraries(hpx::cuda INTERFACE cudart)
   endif()
 
-  target_link_libraries(hpx_base_libraries INTERFACE hpx::cuda)
+  #target_link_libraries(hpx_base_libraries INTERFACE hpx::cuda)
 endif()
