diff --git a/include/hpx/kokkos/deep_copy.hpp b/include/hpx/kokkos/deep_copy.hpp
index 35fa0bc..5ccf191 100644
--- a/include/hpx/kokkos/deep_copy.hpp
+++ b/include/hpx/kokkos/deep_copy.hpp
@@ -14,9 +14,7 @@
 namespace hpx {
 namespace kokkos {
 // TODO: Do we need more overloads here?
-template <typename ExecutionSpace, typename... Args,
-          typename Enable = typename std::enable_if<
-              Kokkos::is_execution_space<ExecutionSpace>::value>::type>
+template <typename ExecutionSpace, typename... Args>
 hpx::future<void> deep_copy_async(ExecutionSpace &&space, Args &&... args) {
   deep_copy(space, std::forward<Args>(args)...);
   return detail::get_future<ExecutionSpace>::call(
